wsgi_kerberos.py

```diff
@@ -49,37 +56,47 @@ def _consume_request(environ, read_max):
     '''
     if read_max == 0:  # Short-circuit early when user opts out of this.
         return
+    if environ["REQUEST_METHOD"] in _NEVER_READ_METHODS:
+        return
     if environ.get("HTTP_EXPECT") == "100-continue":
         return
     try:
-        sock = environ.get('wsgi.input')
-        if hasattr(sock, 'closed') and sock.closed:
+        body = environ.get('wsgi.input')
+        if hasattr(body, 'closed') and body.closed:
             return
 
         content_length = environ.get('CONTENT_LENGTH', '')
         if not content_length:
-            remaining = read_max
-        else:
-            content_length = int(content_length)
-            if content_length > read_max:
-                # User is not willing to read such a large request body, but
-                # reading anything less does not help naively-written clients.
-                return
-            remaining = content_length
-
+            LOG.debug("No Content-Length -> skipping _consume_request")
+            return
+        content_length = int(content_length)
+        if content_length > read_max:
+            # Server is not willing to read such a large request body, but
+            # reading anything less does not help naively-written clients.
+            LOG.warning(
+                "Content-Length (%s) exceeds read_max (%s) -> not consuming"
+                " request body; client may get a connection error. Enabling"
+                " preemptive authentication on the client may avoid this."
+                " You can also pass a higher value of `read_max_on_auth_fail`"
+                " to KerberosAuthMiddleware.",
+                content_length, read_max
+            )
+            return
         # Try to receive all of the data. Keep retrying until we get an error
         # which indicates that we can't retry. Eat errors. The client will just
         # have to deal with a possible Broken Pipe -- we tried.
+        remaining = content_length
         while remaining > 0:
+            to_read = min(io.DEFAULT_BUFFER_SIZE, remaining)
```

Comment:
The default chunk size in werkzeug LimitedStream is 64k - maybe we should use that instead? Using smaller chunk size is much slower for large request bodies.

Reply:
Thanks, I noticed this too and meant to call it out. I did do a little searching when I noticed this, and from what I found, I suspect `io.DEFAULT_BUFFER_SIZE` will often be too small and our default should be increased (see e.g. [this answer](https://stackoverflow.com/a/41747545) to https://stackoverflow.com/questions/2811006/what-is-a-good-buffer-size-for-socket-programming), but I guessed there are cases where 8K could even be too big. I figured exposing this as yet another public API setting to users would be going too far, so I punted and left it set to `io.DEFAULT_BUFFER_SIZE` (which was introduced in Python 2.6, which Werkzeug predates), and had another pang of doubt about whether WSGI-Kerberos is the right layer to be assuming decision-making power about how to consume request bodies when we already known our response.

I updated to match Werkzeug's chunk size in the latest revision.